forc-pkg/src/pkg.rs

```diff
@@ -1957,6 +1962,54 @@ pub fn compile(
         warnings,
         metrics,
     };
+
+    #[cfg(feature = "profile")]
+    {
+        use sway_core::asm_generation::Entry;
+
+        let mut bytes = compiled_package.bytecode.bytes.clone();
+        let data_offset = u64::from_be_bytes(bytes.iter().skip(8).take(8).cloned().collect::<Vec<_>>().try_into().unwrap());
+        let data_section_size = bytes.len() as u64 - data_offset;
+        bytes.truncate(data_offset as usize);
+
+        let mut data_section_used = 0u64;
+
+        fn calculate_entry_size(entry: &Entry) -> u64 {
+            let padding = match &entry.padding {
+                sway_ir::Padding::Left { target_size } => target_size,
+                sway_ir::Padding::Right { target_size } => target_size,
+            };
+
```

Comment:
To future-proof, `std::mem::size_of_val` can be used here instead of magic values

Reply:
Nice, thanks.

Sorry just one last thing - seeing that `sum()` will panic on overflow, it might be worth replacing it with something safer:

```
   - sway_core::asm_generation::Datum::Collection(entries) => entries.iter().map(calculate_entry_size).sum(),
   + sway_core::asm_generation::Datum::Collection(entries) => entries.iter().try_fold(0u64, |acc, &v| acc.checked_add(u64::from(v)));
```

And then deal with a possible error for the `println!()` report.

Reply:
We have taken a look at the data flow for the datum entries, and we have determined that it is the responsibility of the compiler to produce valid data section slot size values that sum within the range of a `u64`. We feel that if this value does overflow, there is an issue in the data section generation code, and a panic might actually be preferred. If the team feels that simply printing a warning would suffice, we can opt for this instead.